### PR TITLE
Add FixedShingleFilter

### PR DESCRIPTION
**Pull request by mocobeta**
Relates to #1204.

Shingle token filter is useful for some advanced applications (e.g., fast phrase search, or implementing word n-gram language models).
Recently I was trying to implement a shingle token filter for my own project. This is more like Lucene's [FixedShingleFilter](https://lucene.apache.org/core/8_10_0/analyzers-common/org/apache/lucene/analysis/shingle/FixedShingleFilter.html) that emits only fixed-sized shingles rather than ShingleFilter; it's easier to implement, and I think fixed-sized shingles are often sufficient in many practical use-cases.

This is my prototype and works well for my application - making shingles is possible by maintaining a buffer that remembers the past tokens (keeps only tokens that are needed to generate the next shingles). However, to me, there is one problem - emitting the correct token position. The current implementation is not compatible with StopWordFilter (or filters that remove some tokens from the stream). `token()` and `token_mut()` methods return a reference rather than a new Token object, so the original stream's position information is not available after overwriting it. I tried to align that problem but have failed so far. But I could be missing something... your help or feedback is welcome.

### CODE REVIEW

**Comment by fulmicoton** [comment]
Here is what I'd love to have for shingles to improve phrase query speed incredibly at a very low cost on indexing.
A shinglefilter that takes a trigger dictionary...

At indexing:
Emit all tokens all of the times
If a given token is encounterred, then also emit the 2-gram starting at the trigger token.

If "and" is a trigger then:

"Of mice and men" would emit at indexing
- "of"{pos=0, length=1}
-  "mice"{pos=1, length=1}
-  "and"{pos=2, length=1}
-  "and+men"{pos=2, length=2} # where + would actually be one these exotic unicode separator codepoint.
- "men"{pos=3, length=1}  

And at query time:
- "of"{pos=0, length=1}
-  "mice"{pos=1, length=1}
-  "and+men"{pos=2, length=2} 


The point here is that the shingle trick is mostly potent for some specific token with a very high doc frequency.
For other words it is not as beneficial to emit shingles, and the cost at indexing is quite high.

**Comment by fulmicoton** [comment]
@mocobeta For this to be useful, we need to have some asymetry between the indexing tokenizer and the search tokenizer isn't it? We need to land #1073 or some variant of it?

**Comment by mocobeta** [comment]
> At indexing:
> Emit all tokens all of the times
> If a given token is encounterred, then also emit the 2-gram starting at the trigger token.

I didn't think of that, but this is an interesting idea to me.
Besides indexing performance, the obvious downside of shingles is their vocabulary size... Generating shingles only when some criteria (dictionary- or rule-based?) are met significantly reduces the index size.

Here are some stats of my tantivy index with bi-gram and tri-gram tokens made by this patch. (data: [47 million web pages](https://lemurproject.org/clueweb09/))
```
Total indexed docs: 47187784
Total term freq in unigram field: 9171726767 (unique terms: 29185)
Total term freq in bigram field: 9124538983 (unique terms: 96691414)
Total term freq in trigram field: 9077354364 (unique terms: 913295070)
```

The whole index size is 100GB (unigram + bi-gram + tri-gram) though, indexing was done within an hour with 36 threads - impressively fast :)

**Comment by mocobeta** [comment]
I think the shingle filter alone works well without #1073, but the combination could be useful in some situations?

File: src/tokenizer/shingle_filter.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    fn push_token(&mut self, mut token: Token) {
+        if self.is_window_filled() {
+            self.pendings.remove(0);
+        }
+        token.position = self.original_token_position;
+        self.pendings.push(token);
+    }
+}
+
+fn make_shingle(token: &mut Token, pendings: &Vec<Token>, buffer: &mut String, separator: char) {
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs]: `&[T]` is almost always a better argument type than `&Vec<T>` .

(that comment does not apply to `&mut Vec<T>`)

```suggestion
fn make_shingle(token: &mut Token, pendings: &[Token], buffer: &mut String, separator: char) {
```

File: src/tokenizer/shingle_filter.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+impl<'a> ShingleFilterStream<'a> {
+    fn is_window_filled(&self) -> bool {
+        self.pendings.len() == self.size - 1
+    }
+
+    fn push_token(&mut self, mut token: Token) {
+        if self.is_window_filled() {
+            self.pendings.remove(0);
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs:48]: nitpick: I wonder if a rolling buffer would be more appropriate here.

File: src/tokenizer/shingle_filter.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+impl<'a> TokenStream for ShingleFilterStream<'a> {
+    fn advance(&mut self) -> bool {
+        while !self.is_window_filled() {
+            if self.tail.advance() {
+                self.push_token(self.tail.token().clone());
+                self.original_token_position += 1;  // This is not compatible with StopWordFilter
+            } else {
+                return false;
+            }
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs]: nitpick:
The whole codebase is not consistent but we try to use early returns these days..

```suggestion
            if !self.tail.advance() {
                return false;
            }
            self.push_token(self.tail.token().clone());
            self.original_token_position += 1;  // This is not compatible with StopWordFilter
```

File: src/tokenizer/shingle_filter.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            make_shingle(
+                &mut self.tail.token_mut(),
+                &self.pendings,
+                &mut self.buffer,
+                self.separator,
+            );
+            self.push_token(next_token);
+            self.original_token_position += 1;
+            return true;
+        }
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs:95]:
```suggestion
        if !self.tail.advance() {
            return false;
        }
        let next_token = self.tail.token().clone();
        make_shingle(
            &mut self.tail.token_mut(),
            &self.pendings,
            &mut self.buffer,
            self.separator,
        );
        self.push_token(next_token);
        self.original_token_position += 1;
        true
   }
```

File: src/tokenizer/shingle_filter.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        false
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs]:
```suggestion
        if !self.tail.advance() {
            return false;
        }
        let next_token = self.tail.token().clone();
        make_shingle(
            &mut self.tail.token_mut(),
            &self.pendings,
            &mut self.buffer,
            self.separator,
        );
        self.push_token(next_token);
        self.original_token_position += 1;
        true
```

File: src/tokenizer/shingle_filter.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            self.pendings.remove(0);
+        }
+        token.position = self.original_token_position;
+        self.pendings.push(token);
+    }
+}
+
+fn make_shingle(token: &mut Token, pendings: &Vec<Token>, buffer: &mut String, separator: char) {
+    buffer.clear();
+    buffer.push_str(token.text.as_str());
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs:57]: nitpick
That just removes one line and one token copy so it does not matter much...
But mem::swap could work too here.

File: src/tokenizer/shingle_filter.rs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        self.tail.token_mut()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::tokenizer::{FixedShingleFilter, SimpleTokenizer, TextAnalyzer, Token};
+
+    #[test]
+    fn test_shingle_bigram() {
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs:112]: The code handles this case just fine but can you add a test where we have less than "window size" tokens to begin with?
(The expected result is that we should emit not token).

Comment (fulmicoton) [reply src/tokenizer/shingle_filter.rs:112]: And one test with the empty string too.

**Review by fulmicoton — CHANGES_REQUESTED** [review]
I put minor comments inline

**Comment by fulmicoton** [comment]
@mocobeta ah yes, you are right. We don't need any asymetry between search and indexing here.
It is only helpful with the extension I suggested.

LGTM.
I just added a few minor comments. Please ignore all of them but the unit tests.

File: src/tokenizer/shingle_filter.rs (hunk 2 above)
Comment (fulmicoton) [reply src/tokenizer/shingle_filter.rs:48]: (please do not modify the code... It is fine as it is)

File: src/tokenizer/shingle_filter.rs (hunk 6 above)
Comment (fulmicoton) [reply src/tokenizer/shingle_filter.rs:57]: (please do not address this comment. the code is fine as is.)

File: src/tokenizer/shingle_filter.rs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        token.position = first_token.position;
+    }
+}
+
+impl<'a> TokenStream for ShingleFilterStream<'a> {
+    fn advance(&mut self) -> bool {
+        while !self.is_window_filled() {
+            if self.tail.advance() {
+                self.push_token(self.tail.token().clone());
+                self.original_token_position += 1;  // This is not compatible with StopWordFilter
```
Comment (fulmicoton) [inline src/tokenizer/shingle_filter.rs]: Can you comment in the code the reason why you chose not to use `self.tail.token().position` for this?
Ideally with an example with a phrase query?

**Review by fulmicoton — CHANGES_REQUESTED** [review]
Added request for comment.

**Comment by fulmicoton** [comment]
```
Total indexed docs: 47187784
Total term freq in unigram field: 9171726767 (unique terms: 29185)
Total term freq in bigram field: 9124538983 (unique terms: 96691414)
Total term freq in trigram field: 9077354364 (unique terms: 913295070)
The whole index size is 100GB (unigram + bi-gram + tri-gram) though, indexing was done within an hour with 36 threads - impressively fast :)
```

Awesome!
So my theory (never tested it) is that the "shingle on trigger" trick gives most of the benefit of shingles, but cuts down the 
number of bigrams drastically.

One mostly wants shingles for super high docfreq terms. In english with 100 static stop words, that would lead to at most 2.9M bigrams (and actually way less than that).

File: src/tokenizer/shingle_filter.rs (hunk 8 above)
Comment (mocobeta) [reply src/tokenizer/shingle_filter.rs]: Thanks for your suggestion. Actually, I still don't have a clear idea about position handling in tantivy; this silly hack should be replaced with more sensible code. I'll try to find the right way for this.

Comment (fulmicoton) [reply src/tokenizer/shingle_filter.rs]: I would not be surprise if we are inconsistent on positions. I would advise you to be very critical on tantivy approach, and I am ok to adopt the lucene way in all tokenizer.

My understanding is that for something like a stop word filter, depending on how we deal with position we will have a different outcome for the following query / doc pair.
- query: the cheapest on internet 
- doc: the cheapest on the internet

In other words, should queries act as 
`the (STOP WORDS)* cheapest (STOP WORDS)* internet`
or 
`the cheapest (STOP WORDS){2} internet`

There is a decent case for both. The best is probably to adopt the lucene way and stay consistent with all TokenFilter

Comment (fulmicoton) [reply src/tokenizer/shingle_filter.rs]: @mocobeta Do you want help to dig into what should be the right choice for position here?

Comment (mocobeta) [reply src/tokenizer/shingle_filter.rs]: @fulmicoton I can't find the time right now, but it could be good homework for the coming winter vacation.
